internal/provider/resource_ably_namespace.go

```diff
@@ -0,0 +1,278 @@
+
+	// Get current state
+	var state AblyNamespace
+	diags = req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Gets the app ID and ID
+	app_id := state.AppID.Value
+	namespace_id := state.ID.Value
+
+	// Instantiates struct of type ably_control_go.Namespace and sets values to output of plan
+	namespace_values := ably_control_go.Namespace{
+		ID:               namespace_id,
+		Authenticated:    plan.Authenticated.Value,
+		Persisted:        plan.Persisted.Value,
+		PersistLast:      plan.PersistLast.Value,
+		PushEnabled:      plan.PushEnabled.Value,
+		TlsOnly:          plan.TlsOnly.Value,
+		ExposeTimeserial: plan.ExposeTimeserial.Value,
+	}
+
+	// Updates an Ably Namespace. The function invokes the Client Library UpdateNamespace method.
+	ably_namespace, err := r.p.client.UpdateNamespace(app_id, &namespace_values)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error updating Resource",
+			"Could not update resource, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	resp_namespaces := AblyNamespace{
+		AppID:            types.String{Value: app_id},
+		ID:               types.String{Value: ably_namespace.ID},
+		Authenticated:    types.Bool{Value: ably_namespace.Authenticated},
+		Persisted:        types.Bool{Value: ably_namespace.Persisted},
+		PersistLast:      types.Bool{Value: ably_namespace.PersistLast},
+		PushEnabled:      types.Bool{Value: ably_namespace.PushEnabled},
+		TlsOnly:          types.Bool{Value: ably_namespace.TlsOnly},
+		ExposeTimeserial: types.Bool{Value: ably_namespace.ExposeTimeserial},
+	}
+
+	// Sets state to new namespace.
+	diags = resp.State.Set(ctx, resp_namespaces)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Delete resource
+func (r resourceNamespace) Delete(ctx context.Context, req tfsdk.DeleteResourceRequest, resp *tfsdk.DeleteResourceResponse) {
+	// Get current state
+	var state AblyNamespace
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Gets the current state. If it is unable to, the provider responds with an error.
+	app_id := state.AppID.Value
+	namespace_id := state.ID.Value
+
+	err := r.p.client.DeleteNamespace(app_id, namespace_id)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error deleting Resource",
+			"Could not delete resource, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	// Remove resource from state
+	resp.State.RemoveResource(ctx)
+}
+
+// Import resource
+func (r resourceNamespace) ImportState(ctx context.Context, req tfsdk.ImportResourceStateRequest, resp *tfsdk.ImportResourceStateResponse) {
+	// Save the import identifier in the id attribute
+	// Recent PR in TF Plugin Framework for paths but Hashicorp examples not updated - https://github.com/hashicorp/terraform-plugin-framework/pull/390
+	tfsdk.ResourceImportStatePassthroughID(ctx, path.Root("id"), req, resp)
```

Comment:
I don't really know what this is or how it works. I assume this will work the same as App as it's just an ID. Though I'm thinking as a namespace is app specific the real id of the namespace is appid+id. A compound key basically. I need to learn how importing works though.